JsonEncoderTest is flaky

It looks like the test round-trips a string to obtain an expected value but fails to request ordered keys like the encoder under test does. As a result, randomly, keys can be in different orders between expected and actual:
```
                     ==================== FAILURES ====================
                     _____ JsonEncoderTest.test_inlined_encoding ______
                     
                     self = <pants_test.engine.test_parsers.JsonEncoderTest testMethod=test_inlined_encoding>
                     relative_json_value = '\n"relative": {\n    "name": "bill",\n    "type_alias": "pants_test.engine.test_parsers.Bob"\n}'
                     is_inline = True
                     
                         def _assert_valid_output(self, relative_json_value, is_inline):
                           # Py3.6+ guarantees order of dictionaries by insertion order, but Py 3.3-3.5 have non-deterministic dictionary
                           # order. So, we first try to assert the order is the same, and then fall back to ignoring order to get the test
                           # working on all interpreters.
                           expected = json.dumps(json.loads(dedent("""
                                 {{
                                   "name": "bob",
                                   {},
                                   "friend": {{
                                     "name": "bill",
                                     "type_alias": "pants_test.engine.test_parsers.Bob"
                                   }},
                                   "type_alias": "pants_test.engine.test_parsers.Bob"
                                 }}
                                 """.format(relative_json_value).strip()
                           )))
                           output = parsers.encode_json(self.bob, inline=is_inline)
                           try:
                     >       self.assertEqual(expected, output)
                     
                     pants_test/engine/test_parsers.py:300: 
                     _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
                     
                     self = <pants_test.engine.test_parsers.JsonEncoderTest testMethod=test_inlined_encoding>
                     first = '{"name": "bob", "relative": {"name": "bill", "type_alias": "pants_test.engine.test_parsers.Bob"}, "type_alias": "pants_test.engine.test_parsers.Bob", "friend": {"name": "bill", "type_alias": "pants_test.engine.test_parsers.Bob"}}'
                     second = '{"relative": {"name": "bill", "type_alias": "pants_test.engine.test_parsers.Bob"}, "name": "bob", "type_alias": "pants_test.engine.test_parsers.Bob", "friend": {"name": "bill", "type_alias": "pants_test.engine.test_parsers.Bob"}}'
                     msg = None
                     
                         def assertEqual(self, first, second, msg=None):
                             """Fail if the two objects are unequal as determined by the '=='
                                    operator.
                                 """
                             assertion_func = self._getAssertEqualityFunc(first, second)
                     >       assertion_func(first, second, msg=msg)
                     
                     /usr/lib/python3.4/unittest/case.py:797: 
                     _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
                     
                     self = <pants_test.engine.test_parsers.JsonEncoderTest testMethod=test_inlined_encoding>
                     first = '{"name": "bob", "relative": {"name": "bill", "type_alias": "pants_test.engine.test_parsers.Bob"}, "type_alias": "pants_test.engine.test_parsers.Bob", "friend": {"name": "bill", "type_alias": "pants_test.engine.test_parsers.Bob"}}'
                     second = '{"relative": {"name": "bill", "type_alias": "pants_test.engine.test_parsers.Bob"}, "name": "bob", "type_alias": "pants_test.engine.test_parsers.Bob", "friend": {"name": "bill", "type_alias": "pants_test.engine.test_parsers.Bob"}}'
                     msg = None
                     
                         def assertMultiLineEqual(self, first, second, msg=None):
                             """Assert that two multi-line strings are equal."""
                             self.assertIsInstance(first, str, 'First argument is not a string')
                             self.assertIsInstance(second, str, 'Second argument is not a string')
                         
                             if first != second:
                                 # don't use difflib if the strings are too long
                                 if (len(first) > self._diffThreshold or
                                     len(second) > self._diffThreshold):
                                     self._baseAssertEqual(first, second, msg)
                                 firstlines = first.splitlines(keepends=True)
                                 secondlines = second.splitlines(keepends=True)
                                 if len(firstlines) == 1 and first.strip('\r\n') == first:
                                     firstlines = [first + '\n']
                                     secondlines = [second + '\n']
                                 standardMsg = '%s != %s' % _common_shorten_repr(first, second)
                                 diff = '\n' + ''.join(difflib.ndiff(firstlines, secondlines))
                                 standardMsg = self._truncateMessage(standardMsg, diff)
                     >           self.fail(self._formatMessage(msg, standardMsg))
                     
                     /usr/lib/python3.4/unittest/case.py:1170: 
                     _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
                     
                     self = <pants_test.engine.test_parsers.JsonEncoderTest testMethod=test_inlined_encoding>
                     msg = '\'{"name": "bob", "relative": {"name": "bill"[182 chars]b"}}\' != \'{"relative": {"name": "bill", "type_alias":[182 c...pants_test.engine.test_parsers.Bob", "friend": {"name": "bill", "type_alias": "pants_test.engine.test_parsers.Bob"}}\n'
                     
                         def fail(self, msg=None):
                             """Fail immediately, with the given message."""
                     >       raise self.failureException(msg)
                     E       AssertionError: '{"name": "bob", "relative": {"name": "bill"[182 chars]b"}}' != '{"relative": {"name": "bill", "type_alias":[182 chars]b"}}'
                     E       - {"name": "bob", "relative": {"name": "bill", "type_alias": "pants_test.engine.test_parsers.Bob"}, "type_alias": "pants_test.engine.test_parsers.Bob", "friend": {"name": "bill", "type_alias": "pants_test.engine.test_parsers.Bob"}}
                     E       + {"relative": {"name": "bill", "type_alias": "pants_test.engine.test_parsers.Bob"}, "name": "bob", "type_alias": "pants_test.engine.test_parsers.Bob", "friend": {"name": "bill", "type_alias": "pants_test.engine.test_parsers.Bob"}}
                     
                     /usr/lib/python3.4/unittest/case.py:642: AssertionError
                     
                     During handling of the above exception, another exception occurred:
                     
                     self = <pants_test.engine.test_parsers.JsonEncoderTest testMethod=test_inlined_encoding>
                     
                         def test_inlined_encoding(self):
                           relative_json_value = dedent('''
                               "relative": {
                                   "name": "bill",
                                   "type_alias": "pants_test.engine.test_parsers.Bob"
                               }''')
                     >     self._assert_valid_output(relative_json_value=relative_json_value, is_inline=True)
                     
                     pants_test/engine/test_parsers.py:319: 
                     _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
                     pants_test/engine/test_parsers.py:307: in _assert_valid_output
                         self.assertEqual(convert_to_stripped_set(expected), convert_to_stripped_set(output))
                     E   AssertionError: Items in the first set but not the second:
                     E   '"name": "bob"'
                     E   ' "relative": "name": "bill"'
                     E   Items in the second set but not the first:
                     E   '"relative": "name": "bill"'
                     E   ' "name": "bob"'
                     _____ JsonEncoderTest.test_shallow_encoding ______
                     
                     self = <pants_test.engine.test_parsers.JsonEncoderTest testMethod=test_shallow_encoding>
                     relative_json_value = '"relative": "::an opaque address::"'
                     is_inline = False
                     
                         def _assert_valid_output(self, relative_json_value, is_inline):
                           # Py3.6+ guarantees order of dictionaries by insertion order, but Py 3.3-3.5 have non-deterministic dictionary
                           # order. So, we first try to assert the order is the same, and then fall back to ignoring order to get the test
                           # working on all interpreters.
                           expected = json.dumps(json.loads(dedent("""
                                 {{
                                   "name": "bob",
                                   {},
                                   "friend": {{
                                     "name": "bill",
                                     "type_alias": "pants_test.engine.test_parsers.Bob"
                                   }},
                                   "type_alias": "pants_test.engine.test_parsers.Bob"
                                 }}
                                 """.format(relative_json_value).strip()
                           )))
                           output = parsers.encode_json(self.bob, inline=is_inline)
                           try:
                     >       self.assertEqual(expected, output)
                     
                     pants_test/engine/test_parsers.py:300: 
                     _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
                     
                     self = <pants_test.engine.test_parsers.JsonEncoderTest testMethod=test_shallow_encoding>
                     first = '{"name": "bob", "relative": "::an opaque address::", "type_alias": "pants_test.engine.test_parsers.Bob", "friend": {"name": "bill", "type_alias": "pants_test.engine.test_parsers.Bob"}}'
                     second = '{"relative": "::an opaque address::", "name": "bob", "type_alias": "pants_test.engine.test_parsers.Bob", "friend": {"name": "bill", "type_alias": "pants_test.engine.test_parsers.Bob"}}'
                     msg = None
                     
                         def assertEqual(self, first, second, msg=None):
                             """Fail if the two objects are unequal as determined by the '=='
                                    operator.
                                 """
                             assertion_func = self._getAssertEqualityFunc(first, second)
                     >       assertion_func(first, second, msg=msg)
                     
                     /usr/lib/python3.4/unittest/case.py:797: 
                     _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
                     
                     self = <pants_test.engine.test_parsers.JsonEncoderTest testMethod=test_shallow_encoding>
                     first = '{"name": "bob", "relative": "::an opaque address::", "type_alias": "pants_test.engine.test_parsers.Bob", "friend": {"name": "bill", "type_alias": "pants_test.engine.test_parsers.Bob"}}'
                     second = '{"relative": "::an opaque address::", "name": "bob", "type_alias": "pants_test.engine.test_parsers.Bob", "friend": {"name": "bill", "type_alias": "pants_test.engine.test_parsers.Bob"}}'
                     msg = None
                     
                         def assertMultiLineEqual(self, first, second, msg=None):
                             """Assert that two multi-line strings are equal."""
                             self.assertIsInstance(first, str, 'First argument is not a string')
                             self.assertIsInstance(second, str, 'Second argument is not a string')
                         
                             if first != second:
                                 # don't use difflib if the strings are too long
                                 if (len(first) > self._diffThreshold or
                                     len(second) > self._diffThreshold):
                                     self._baseAssertEqual(first, second, msg)
                                 firstlines = first.splitlines(keepends=True)
                                 secondlines = second.splitlines(keepends=True)
                                 if len(firstlines) == 1 and first.strip('\r\n') == first:
                                     firstlines = [first + '\n']
                                     secondlines = [second + '\n']
                                 standardMsg = '%s != %s' % _common_shorten_repr(first, second)
                                 diff = '\n' + ''.join(difflib.ndiff(firstlines, secondlines))
                                 standardMsg = self._truncateMessage(standardMsg, diff)
                     >           self.fail(self._formatMessage(msg, standardMsg))
                     
                     /usr/lib/python3.4/unittest/case.py:1170: 
                     _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
                     
                     self = <pants_test.engine.test_parsers.JsonEncoderTest testMethod=test_shallow_encoding>
                     msg = '\'{"name": "bob", "relative": "::an opaque ad[137 chars]b"}}\' != \'{"relative": "::an opaque address::", "name[137 c..."bill", "type_alias": "pants_test.engine.test_parsers.Bob"}}\n?                                      +++++++++++++++\n'
                     
                         def fail(self, msg=None):
                             """Fail immediately, with the given message."""
                     >       raise self.failureException(msg)
                     E       AssertionError: '{"name": "bob", "relative": "::an opaque ad[137 chars]b"}}' != '{"relative": "::an opaque address::", "name[137 chars]b"}}'
                     E       - {"name": "bob", "relative": "::an opaque address::", "type_alias": "pants_test.engine.test_parsers.Bob", "friend": {"name": "bill", "type_alias": "pants_test.engine.test_parsers.Bob"}}
                     E       ?  ---------------
                     E       + {"relative": "::an opaque address::", "name": "bob", "type_alias": "pants_test.engine.test_parsers.Bob", "friend": {"name": "bill", "type_alias": "pants_test.engine.test_parsers.Bob"}}
                     E       ?                                      +++++++++++++++
                     
                     /usr/lib/python3.4/unittest/case.py:642: AssertionError
                     
                     During handling of the above exception, another exception occurred:
                     
                     self = <pants_test.engine.test_parsers.JsonEncoderTest testMethod=test_shallow_encoding>
                     
                         def test_shallow_encoding(self):
                           relative_json_value = '"relative": "::an opaque address::"'
                     >     self._assert_valid_output(relative_json_value=relative_json_value, is_inline=False)
                     
                     pants_test/engine/test_parsers.py:311: 
                     _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
                     pants_test/engine/test_parsers.py:307: in _assert_valid_output
                         self.assertEqual(convert_to_stripped_set(expected), convert_to_stripped_set(output))
                     E   AssertionError: Items in the first set but not the second:
                     E   ' "relative": "::an opaque address::"'
                     E   '"name": "bob"'
                     E   Items in the second set but not the first:
                     E   '"relative": "::an opaque address::"'
                     E   ' "name": "bob"'
                      generated xml file: /home/travis/build/pantsbuild/pants/.pants.d/test/pytest/tests.python.pants_test.engine.parsers/junitxml/TEST-tests.python.pants_test.engine.parsers.xml 
                     ============ slowest 3 test durations ============
                     0.00s call     pyprep/sources/39575ac66d65de953240da57d5fb6304a539b58f/tests/python/pants_test/engine/test_parsers.py::JsonEncoderTest::test_shallow_encoding
                     0.00s call     pyprep/sources/39575ac66d65de953240da57d5fb6304a539b58f/tests/python/pants_test/engine/test_parsers.py::JsonEncoderTest::test_inlined_encoding
                     0.00s call     pyprep/sources/39575ac66d65de953240da57d5fb6304a539b58f/tests/python/pants_test/engine/test_parsers.py::JsonParserTest::test_error_presentation
                     ====== 2 failed, 12 passed in 0.25 seconds =======
```
Seen here: https://travis-ci.org/pantsbuild/pants/jobs/424559912

--- a/tests/python/pants_test/engine/test_parsers.py
+++ b/tests/python/pants_test/engine/test_parsers.py
@@ -279,47 +279,38 @@ class JsonEncoderTest(unittest.TestCase):
 
     self.bob = Bob(name='bob', relative=resolvable_bill, friend=bill)
 
-  def _assert_valid_output(self, relative_json_value, is_inline):
-    # Py3.6+ guarantees order of dictionaries by insertion order, but Py 3.3-3.5 have non-deterministic dictionary
-    # order. So, we first try to assert the order is the same, and then fall back to ignoring order to get the test
-    # working on all interpreters.
-    expected = json.dumps(json.loads(dedent("""
-        {{
-          "name": "bob",
-          {},
-          "friend": {{
-            "name": "bill",
-            "type_alias": "pants_test.engine.test_parsers.Bob"
-          }},
-          "type_alias": "pants_test.engine.test_parsers.Bob"
-        }}
-        """.format(relative_json_value).strip()
-    )))
-    output = parsers.encode_json(self.bob, inline=is_inline)
-    try:
-      self.assertEqual(expected, output)
-    except AssertionError:
-      def convert_to_stripped_set(json_str):
-        no_brackets = json_str.replace('{', '').replace('}', '')
-        # TODO: #6275 appears to have broken this so that the `.strip()` is required now, but
-        # it's not clear why (why would there suddenly be extra whitespace on either side of each
-        # line, for multiple tests, where there wasn't before?).
-        distinct_lines = set(l.strip() for l in no_brackets.split(','))
-        return distinct_lines
-
-      self.assertEqual(convert_to_stripped_set(expected), convert_to_stripped_set(output))
-
   def test_shallow_encoding(self):
-    relative_json_value = '"relative": "::an opaque address::"'
-    self._assert_valid_output(relative_json_value=relative_json_value, is_inline=False)
+    expected_json = dedent("""
+    {
+      "name": "bob",
+      "type_alias": "pants_test.engine.test_parsers.Bob",
+      "friend": {
+        "name": "bill",
+        "type_alias": "pants_test.engine.test_parsers.Bob"
+      },
+      "relative": "::an opaque address::"
+    }
+    """).strip()
+    self.assertEqual(json.dumps(json.loads(expected_json), sort_keys=True),
+                     parsers.encode_json(self.bob, inline=False, sort_keys=True))
 
   def test_inlined_encoding(self):
-    relative_json_value = dedent('''
+    expected_json = dedent("""
+    {
+      "name": "bob",
+      "type_alias": "pants_test.engine.test_parsers.Bob",
+      "friend": {
+        "name": "bill",
+        "type_alias": "pants_test.engine.test_parsers.Bob"
+      },
       "relative": {
-          "name": "bill",
-          "type_alias": "pants_test.engine.test_parsers.Bob"
-      }''')
-    self._assert_valid_output(relative_json_value=relative_json_value, is_inline=True)
+        "name": "bill",
+        "type_alias": "pants_test.engine.test_parsers.Bob"
+      }
+    }
+    """).strip()
+    self.assertEqual(json.dumps(json.loads(expected_json), sort_keys=True),
+                     parsers.encode_json(self.bob, inline=True, sort_keys=True))
 
 
 class PythonAssignmentsParserTest(unittest.TestCase):

